### PR TITLE
Make weight calculator work for SubscriptionLineItems

### LINKED ISSUE #4410: Subs cannot be edited
## Description

When a manager or super admin tries to edit a subscription, the changes are not taken into account.

## Steps to Reproduce

1. Got to `/admin/subscriptions`
2. Create a new subs, and edit it or click straight on the edit button of a subs already existing
3. Change something e.g. add a product or change a quantity. Save
4. The display shows the new info
5. Now come back to `/admin/subscriptions` and check your subscription again. You will see nothing has been actually saved.

## Workaround
The workaround is to cancel the subs and create a new one.

## Severity
bug-s3: a feature is broken but there is a workaround

## Your Environment
* Version used: v2.5.3
* Browser name and version: Firefox 70
* Operating System and version (desktop or mobile): Desktop

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #4410 and #4322 

Make weight calculator work for SubscriptionLineItems.
We do this by making the calculator test if line_item responds to final_weight_volume field. See commit messages for more details.

#### What should we test?
Create subscription with a shipping method with a weight based calculator. Make sure you can now edit the subscription in both the subs list and the edit subs page. Make sure you can paused/unpaused the subscription.

#### Release notes
Changelog Category: Fixed
Fixed bug in subscriptions with shipping methods with price based on weight. These subscriptions can now be edited and paused/unpaused.

### CODE REVIEW

File: app/models/calculator/weight.rb [hunk 1]
```diff
@@ -25,21 +25,26 @@ def total_weight(line_items)
...
+      end
     end
 
     def weight_per_final_weight_volume(line_item)
-      if line_item.variant.product.andand.variant_unit == 'weight'
-        # Divided by 1000 because grams is the base weight unit and the calculator price is per_kg
-        line_item.final_weight_volume / 1000.0
+      if variant_unit(line_item) == 'weight'
+        # The calculator price is per_kg so we need to convert final_weight_volume to kg
+        convert_g_to_kg(line_item.final_weight_volume)
```
Comment (sauloperez) [inline app/models/calculator/weight.rb]: :ok_hand:

**Comment by sauloperez** [comment]
Didn't we say in https://github.com/openfoodfoundation/openfoodnetwork/issues/4322#issuecomment-546399235 we needed to implement `SubscriptionLineItem#final_weight_volume`?

**Comment by luisramos0** [comment]
Hello Pau, yes we did but I changed my mind after analysis. This PR is what I suggest we do: make weight calculator handle line_items without final_weight_volume. This way we dont need to replicate logic for SubscriptionLineItems and all the weight logic stays in this calculator.

This logic, needed for SubscriptionLineItems, also applies to line_items without final_weight_volume with variant_unit weight. This way we can relax about final_weigth_volume being mandatory for this calculation. It makes the model more simple with one less restriction and the calculator more robust.

**Comment by RachL** [comment]
@luisramos0 I've tested subs with the wieght calculator in shipping and without and I can both edit and unpause :tada:

In terms of automated tests, do we need to introduce some tests during release testing on this or are we now covered?

**Comment by luisramos0** [comment]
I think we are good in terms of testing. the new unit test makes sure the error is shown in the UI and the new functional test validates there is no error.

**Comment by sauloperez** [comment]
> In terms of automated tests, do we need to introduce some tests during release testing on this or are we now covered?

@RachL I was as excited as you about it :joy: . I think this means we can stop testing subs in releases. That's what we last spoke, right? I guess we need to confirm with test coverage.

**Pull request merged**